### PR TITLE
feat: add Gitlab PR generator webhooks support

### LINKED ISSUE #10234: ApplicationSet Gitlab PR generator hooks support
# Summary

Currently Gitlab webhooks have no effect on ApplicationSet PR Generator.

# Motivation

This will improve the user experience as updates will be delivered faster. We'll also get consistency between gitlab and github behavior.

# Proposal

As far as I understand, this is due to the fact that the _getPRGeneratorInfo_ function does not support Gitlab events
https://github.com/argoproj/argo-cd/blob/b8ad57f2d421a365707cfa62c64c3644a0e4c533/applicationset/utils/webhook.go#L180

It's looks like it should handle Gitlab Merge request events https://docs.gitlab.com/ee/user/project/integrations/webhook_events.html#merge-request-events

### PR DESCRIPTION
**Pull request by Enity**
Closes [#10234](https://github.com/argoproj/argo-cd/issues/10234)

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

### CODE REVIEW

**Comment by Enity** [comment]
Tested manually on my own cluster. Works well. 
Docker image from this PR: https://hub.docker.com/r/enity/argocd-gitlab-mr-hooks

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]
Almost perfect! One nitpick, plus, would you mind compressing the image? squoosh.app gets it down to 60kb from 600kb if I save it as a JPEG (which I think is fine for this use case).

File: applicationset/utils/webhook.go [hunk 1]
```diff
@@ -266,25 +308,53 @@ func shouldRefreshPRGenerator(gen *v1alpha1.PullRequestGenerator, info *prGenera
...
+		return true
 	}
-	if !info.Github.APIRegexp.MatchString(api) {
-		log.Debugf("%s does not match %s", gen.Github.API, info.Github.APIRegexp.String())
-		return false
+
+	if gen.Github != nil && info.Github != nil {
+		if gen.Github == nil || info.Github == nil {
+			return false
+		}
```
[the remark this answers is not in the document]
Comment (Enity) [reply applicationset/utils/webhook.go]: You are right, i missed it. Fixed

**Comment by Enity** [comment]
> Almost perfect! One nitpick, plus, would you mind compressing the image? squoosh.app gets it down to 60kb from 600kb if I save it as a JPEG (which I think is fine for this use case).

Picture is now 111kb, i kept it png just for consistency. Don't know why but all pictures is png.

File: applicationset/utils/webhook.go [hunk 2]
```diff
@@ -266,25 +308,50 @@ func shouldRefreshPRGenerator(gen *v1alpha1.PullRequestGenerator, info *prGenera
...
+		if gen.Github.Repo != info.Github.Repo {
+			return false
+		}
+		api := gen.Github.API
+		if api == "" {
+			api = "https://api.github.com/"
+		}
+		if !info.Github.APIRegexp.MatchString(api) {
+			log.Debugf("%s does not match %s", api, info.Github.APIRegexp.String())
```
Comment (Enity) [inline applicationset/utils/webhook.go:347]: Also i found a little bug. This line should log **api** variable, not **gen.Github.API**

Comment (Enity) [reply applicationset/utils/webhook.go:347]: already fixed it

File: applicationset/utils/webhook.go [hunk 3]
```diff
@@ -202,15 +209,31 @@ func getPRGeneratorInfo(payload interface{}) *prGeneratorInfo {
...
+
+		apiURL := payload.Project.WebURL
+		urlObj, err := url.Parse(apiURL)
+		if err != nil {
+			log.Errorf("Failed to parse repoURL '%s'", apiURL)
+			return nil
+		}
+
+		info.Gitlab = &prGeneratorGitlabInfo{
+			Project:     strconv.FormatInt(payload.ObjectAttributes.TargetProjectID, 10),
```
Comment (Enity) [inline applicationset/utils/webhook.go:225]: Does it good solution for int cast? sorry, I'm not a go developer 😀

File: applicationset/utils/webhook.go (hunk 2 above)
Comment (crenshaw-dev) [reply applicationset/utils/webhook.go:347]: Nice!

File: applicationset/utils/webhook.go (hunk 3 above)
Comment (crenshaw-dev) [reply applicationset/utils/webhook.go:225]: I think this is fine!

**Review by crenshaw-dev — APPROVED** [review]
Thanks @Enity!

**Pull request merged**